text/0000-generic-int.md

```diff
@@ -0,0 +1,552 @@
+impl MyDefault for usize {
+    fn my_default() -> usize { 1 }
+}
+impl MyDefault for isize {
+    fn my_default() -> isize { 1 }
+}
+```
+
+That's a lot better! As you'll notice, we still have to separately implement
+code for `usize` and `isize`; that's because they're separate types altogether.
+For example, on 64-bit systems, `usize` is technically just a `u64`, but
+`uint<64>` refers to `u64`, not `usize`. It's unfortunate, but we've reduced our
+number of impls from twelve to four, which is still a very big deal!
+
+One other side effect of having `uint<N>` and `int<N>` is that we can represent
+a lot more types than before. For example, `uint<7>` is just a seven-bit
+integer, which we might use to represent an ASCII character. Don't be fooled by
+these; we can't actually make `N` bigger than 128 (for now), and `uint<N>` is
+almost always stored using `uint<N.next_power_of_two()>`. `uint<7>` is stored as
+`u8`, `uint<10>` is stored as `u16`, `uint<48>` is stored as `u64`, etc. If you
+store a value with a nonstandard amount of bits, you're not actually saving any
+space, so, keep that in mind!
+
+That said, there are a few cases where we might want to use one of these
+weirdly sized types, as many data formats store values in terms of bits instead
+of bytes to save space. Eventually, we'll have a way of representing these, but
+not today.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Primitive behaviour
+
+The compiler will have two new built-in integer types: `uint<N>` and `int<N>`,
+where `const N: usize`. These will alias to existing `uN` and `iN` types if `N`
+is a power of two and no greater than 128. `usize` and `isize` remain separate
+types due to coherence issues, and `bool` remains separate from `uint<1>` as it
+does not have most of the functionality that integers have.
+
+`int<N>` and `uint<N>` will have the smallest power-of-two size and alignment.
+For example, this means that `uint<48>` will take up 8 bytes and have an
+alignment of 8, even though it only has 6 bytes of data.
+
+`int<N>` store values between -2<sup>N-1</sup> and 2<sup>N-1</sup>-1, and
+`uint<N>` stores values between 0 and 2<sup>N</sup>-1. One unexpected case of
+this is that `i1` represents zero or *negative* one, even though LLVM and other
+places use `i1` to refer to `u1`. This case is left as-is because generic code
+*should* expect sign extension to happen for all signed integers, and the
+`N = 1` case is no different. It's valid to cast `int<N>` or `uint<N>` to
+`int<M>` or `uint<M>` via `as`, and sign extension will occur as expected. This
+conversion is lossless when converting `int<N>` to `int<M>` or `uint<M>` or
+from `uint<N>` to `int<M>` or `uint<M + 1>`, where `M >= N`.
+
+In addition to the usual casts, `u1` and `i1` can also be cast *to* `bool` via
+`as`, whereas most integer types can only be cast from `bool`.
+
+For the moment, a monomorphisation error will occur if `N > 128`, to minimise
```

Comment:
This is my biggest concern; I don't think monomorphization of this magnitude errors belong in the language and I would like to see this changed *before* stabilization.

Reply:
Would you be willing to elaborate more on this? Why are these errors a problem?

Reply:
The nature of Rust's bounded polymorphism is that type checking should be modular so that you can check a generic function separately from the instantiation and then the instantiation of any parameters satisfying bounds should not result in errors.

This makes for more local error messages (you won't have post monomorphization errors in some location far removed from where instantiation happened...), fewer surprises (because this is how polymorphism works everywhere else in the type system) and possibly also better performance.
Another benefit of avoiding post monomorphization errors is that the need to monomorphize as an implementation strategy is lessened. That said, there are instances where the compiler will cause post monomorphization errors, but those are extremely unlikely to occur in actual code. In the case of `N > 128` it is rather quite likely.

The general principle is that you declare up front the requirements (with bounds, etc.) to use / call an object and then you don't impose new and hidden requirements for certain values.

If you want to impose `N > 128`, then that should be explicitly required in the "signatures", e.g. you should state `struct uint<const N: usize> where N <= 128 { .. }` (and on the impls...). Otherwise, it should work for all `N: usize` evaluable at compile time.

Reply:
> e.g. you should state `struct uint<const N: usize> where N <= 128 { .. }` (and on the impls...)

Is this possible with any currently planned or near future const generics? I don't recall seeing any RFCs that would support const operations in where clauses (although I would love to have them available). This specific case might be possible with some horrible hack like

```rust
trait LessThan128 {}
struct IsLessThan128<const N: usize>;

impl LessThan128 for IsLessThan128<0> {}
⋮
impl LessThan128 for IsLessThan128<127> {}

struct uint<const N: usize> where IsLessThan128<N>: LessThan128 { .. }
```

but :nauseated_face:

Reply:
@Nemo157 not possible with RFC 2000 but might be with future extensions.

Reply:
This is a different formulation which might work(if you can use a const expression in associated types):

struct Bool<const B:bool>;

trait LessThan128<const N: usize> {
    type IsIt;
}

impl<const N:usize> LessThan128<N> for () {
    type IsIt=Bool<{N<128}>;
}

struct uint<const N: usize> 
where (): LessThan128<N,Output=Bool<true>> 
{ .. }

Reply:
AFAIK stable Rust does not currently have any monomorphization time errors - if you find any, it's a bug - so this RFC "as is" would be introducing them into the language =/

Reply:
@gnzlbg 

```rust
fn poly<T>() {
    let _n: [T; 10000000000000];
}

fn monomorphization_error() {
    poly::<u8>(); // OK!
    poly::<String>(); // BOOM!
}

fn main() {
    monomorphization_error();
}
```